lib/luaxtable.c

```diff
@@ -43,37 +47,123 @@ typedef struct luaxtable_s {
 		struct xt_match match;
 		struct xt_target target;
 	};
+	int fallback;
 	luaxtable_type_t type;
 } luaxtable_t;
 
-static bool luaxtable_match(const struct sk_buff *skb, struct xt_action_param *par)
-{
-	return true;
-}
+static lunatik_object_t *xtable_targethooks = NULL;
+static lunatik_object_t *xtable_matchhooks = NULL;
 
-static int luaxtable_match_check(const struct xt_mtchk_param *par)
+static inline void luaxtable_newdata(lua_State *L, void *ptr, size_t size, bool sleep)
 {
-	return 0;
+	lunatik_require(L, data);
+	lunatik_object_t *data = lunatik_checknull(L, luadata_new(ptr, size, sleep));
+	lunatik_cloneobject(L, data);
 }
 
-static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
+#define luaxtable_getrcu(xtable, hook, key, keylen)			\
+do {									\
+	lunatik_object_t *obj = luarcu_gettable(xtable_##hook##hooks, key, keylen);	\
+	if (obj == NULL) {						\
+		pr_err("could not find table for %s\n", key);		\
+		return -1;						\
+	}								\
+	xtable = (luaxtable_t *)obj->private;				\
+} while(0)
+
+#define luaxtable_setrcu(object, hook, key, keylen)			\
+do {									\
+	if (luarcu_settable(xtable_##hook##hooks, key, keylen, object) != 0) {	\
+		luaL_error(L, "unable to set table");			\
+		return -1;						\
+	}								\
+} while(0)
+
+
+static int luaxtable_invoke(lua_State *L, luaxtable_t *xtable, struct sk_buff *skb, const char *op, int nret)
 {
+	int base = lua_gettop(L);
+	int ret = -ENXIO;
+	int nargs = 0;
+
+	if (lunatik_getregistry(L, xtable) != LUA_TTABLE) {
+		pr_err("could not find ops table\n");
+		goto err;
+	}
+
+	if (lua_getfield(L, -1, op) != LUA_TFUNCTION) {
+		pr_err("%s isn't defined\n", op);
+		goto err;
+	}
+
+	if (skb != NULL ){
+		luaxtable_newdata(L, skb, sizeof(struct sk_buff), false);
+		nargs = 1;
+	}
+
+	if (lua_pcall(L, nargs, nret, 0) != LUA_OK) {
+		pr_err("%s error: %s\n", op, lua_tostring(L, -1));
+		ret = -ECANCELED;
+		goto err;
+	}
+
+	return nret > 0 ? luaL_optinteger(L, -1, 0) : 0;
+err:
+	lua_settop(L, base); /* pop everything, including args */
+	return ret;
 }
 
-static unsigned int luaxtable_target(struct sk_buff *skb, const struct xt_action_param *par)
-{
-	return NF_ACCEPT;
+#define LUAXTABLE_HOOK_CB(T, U, V, hook, huk) 				\
+static T luaxtable_##hook(U skb, V par)					\
+{									\
+	const luaxtable_info_t *info = (const luaxtable_info_t *)par->huk##info;	\
+	luaxtable_t *xtable = info->data;				\
+	int ret = xtable->fallback;					\
+									\
+	/* TODO: handle const type in luadata */			\
+	lunatik_run(xtable->runtime, luaxtable_invoke, ret, xtable, (struct sk_buff *)skb, #hook, 1);	\
+	if (ret < 0)							\
+		pr_err("error in running " #hook " lua handler : %d\n", ret);	\
+	return ret < 0 ? xtable->fallback : ret;					\
```

Comment:
```suggestion
	return ret;					\
```

Reply:
Here, `ret` is being changed by lunatik_run

Reply:
which is set by your `invoke` function, doesn't it? why you need to check it again?

Reply:
I need to differentiate between negative returns from pcall and an error condition during or before pcall in `invoke`. If the return is negative from the pcall result, it can be returned, but for errors we need to have a fallback value.

Reply:
Ah, I just remembered, `match` should return bool and `target` should return `unsigned int`, so we can't return negative values, So only for these callbacks we need to send that fallback value. For checkentry, a negative return is fine (a success is 0) and for destroy no value is to be returned.